boxtree/tree_build.py

```diff
@@ -46,6 +56,240 @@ class MaxLevelsExceeded(RuntimeError):
+        coarsen_parents = tob.box_parent_ids[coarsen_sources]
+        coarsen_peers = tob.box_child_ids[:, coarsen_parents]
+        coarsen_peer_is_leaf = tob.get_leaf_flags()[coarsen_peers]
+        coarsen_exec_flags = np.all(coarsen_peer_is_leaf, axis=0)
+
+        coarsen_flags_ignored = (coarsen_exec_flags != coarsen_flags)
+        if np.any(coarsen_flags_ignored):
+            import warnings
+            warnings.warn(f"{np.sum(coarsen_flags_ignored)} out of "
+                          f"{np.sum(coarsen_flags)} coarsening flags ignored "
+                          "to prevent removing non-leaf boxes")
+
+        coarsen_parents = coarsen_parents[coarsen_exec_flags]
+        coarsen_peers = coarsen_peers[:, coarsen_exec_flags]
+        box_parents[coarsen_peers] = -1
+        box_children[coarsen_parents] = 0
+
+    return TreeOfBoxes(
+            box_centers=box_centers,
+            root_extent=tob.root_extent,
+            box_parent_ids=box_parents,
+            box_child_ids=box_children,
+            box_levels=box_levels)
+
+
+# }}} End refine/coarsen
+
+
+def make_tob_root(dim, bbox):
+    """
+    Make the minimal tree of boxes.
+    """
+    center = np.array(
+        [(bbox[0][iaxis] + bbox[1][iaxis]) * 0.5 for iaxis in range(dim)])
+    extents = np.array(
+        [(bbox[1][iaxis] - bbox[0][iaxis]) for iaxis in range(dim)])
+    from pytools import single_valued
+    parents = np.array([0], np.intp)
+    parents[0] = -1  # root has no parent
+    return TreeOfBoxes(
+            box_centers=center.reshape(dim, 1),
+            root_extent=single_valued(extents, np.allclose),
+            box_parent_ids=parents,
+            box_child_ids=np.array([0] * 2**dim, np.intp).reshape(2**dim, 1),
+            box_levels=np.array([0]),
+            )
+
+
+def make_mesh_from_leaves(tob):
+    """Make a :mod:`meshmode` mesh from the leaf boxes.
+
+    :arg tob: a :class:`TreeOfBoxes`.
+    """
+    lfboxes = tob.leaf_boxes()
+    lfcenters = tob.box_centers[:, lfboxes]
+    lflevels = tob.box_levels[lfboxes]
+    lfradii = tob.root_extent / 2 / (2**lflevels)
+
+    # use tensor product nodes ordering
+    import modepy.nodes as nd
+    cell_nodes_1d = np.array([-1, 1])
+    cell_nodes = nd.tensor_product_nodes(tob.dim, cell_nodes_1d)
+
+    lfvertices = (
+        np.repeat(lfcenters, 2**tob.dim, axis=1)
+        + np.repeat(lfradii, 2**tob.dim) * np.tile(cell_nodes, (1, len(lfboxes)))
+    )
+
+    # FIXME: purge redundant vertices
+    from meshmode.mesh import Mesh, TensorProductElementGroup
+    from meshmode.mesh.generation import make_group_from_vertices
+
+    vertex_indices = np.arange(
+        len(lfboxes) * 2**tob.dim, dtype=np.int32).reshape([2**tob.dim, -1]).T
+    group = make_group_from_vertices(
+        lfvertices, vertex_indices, 1,
+        group_cls=TensorProductElementGroup,
+        unit_nodes=None)
+
+    return Mesh(vertices=lfvertices, groups=[group])
+
+
+def make_global_leaf_quadrature(actx, tob, order):
```

Comment:
This feels overly specific, given that it picks a quadrature on the users behalf. Maybe move this function to the tests? Or underscore it to mark it internal?

Reply:
a609966